File: source-gallery/src/main/java/com/google/android/apps/muzei/gallery/GalleryImportPhotosDialogFragment.java
package com.google.android.apps.muzei.gallery;

import android.app.Dialog;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.widget.ArrayAdapter;

import java.util.ArrayList;
import java.util.List;

public class GalleryImportPhotosDialogFragment extends DialogFragment {
    public static final String TAG = "GalleryImportPhotosDialogFragment";

    private LiveData<List<ActivityInfo>> mGetContentActivitiesLiveData;
    private OnRequestContentListener mListener;
    private ArrayAdapter<CharSequence> mAdapter;

    @NonNull
    public static GalleryImportPhotosDialogFragment createInstance() {
        return new GalleryImportPhotosDialogFragment();
    }

    public GalleryImportPhotosDialogFragment () {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final TypedArray a = getContext().obtainStyledAttributes(null, R.styleable.AlertDialog,
                R.attr.alertDialogStyle, 0);
        @LayoutRes int listItemLayout = a.getResourceId(R.styleable.AlertDialog_listItemLayout, 0);
        a.recycle();
        mAdapter = new ArrayAdapter<>(getContext(), listItemLayout);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new AlertDialog.Builder(getContext())
                .setTitle(R.string.gallery_import_dialog_title)
                .setAdapter(mAdapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        List<ActivityInfo> getContentActivites = mGetContentActivitiesLiveData.getValue();
                        if (getContentActivites != null) {
                            mListener.requestGetContent(mGetContentActivitiesLiveData.getValue().get(which));
                        }
                    }
                }).create();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnRequestContentListener) {
            mListener = (OnRequestContentListener) context;
        } else {
            throw new IllegalArgumentException(context.getClass().getSimpleName() + " must implement OnRequestContentListener");
        }
        mGetContentActivitiesLiveData = ViewModelProviders.of(getActivity())
                .get(GallerySettingsViewModel.class)
                .getGetContentActivityInfoList();
        mGetContentActivitiesLiveData.observe(this, new Observer<List<ActivityInfo>>() {
            @Override
            public void onChanged(@Nullable List<ActivityInfo> getContentActivites) {
                if (getContentActivites == null || getContentActivites.isEmpty()) {
                    dismiss();
                    return;
                }
                updateAdapter(getContentActivites);
            }
        });
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    private void updateAdapter(@NonNull List<ActivityInfo> getContentActivites) {
        PackageManager packageManager = getContext().getPackageManager();
        int size = getContentActivites.size();
        List<CharSequence> items = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            items.add(getContentActivites.get(i).loadLabel(packageManager));
        }
        mAdapter.clear();
        mAdapter.addAll(items);
        mAdapter.notifyDataSetChanged();
    }

    public interface OnRequestContentListener {
        void requestGetContent(ActivityInfo info);
    }
}

File: source-gallery/src/main/java/com/google/android/apps/muzei/gallery/GalleryImportPhotosDialogFragment.kt
package com.google.android.apps.muzei.gallery

import android.app.Dialog
import android.content.Context
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.widget.ArrayAdapter
import androidx.annotation.LayoutRes
import androidx.core.content.withStyledAttributes
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.viewModels
import androidx.lifecycle.LiveData
import androidx.lifecycle.observe
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class GalleryImportPhotosDialogFragment : DialogFragment() {

    companion object {
        private const val TAG = "GalleryImportPhotosDialogFragment"

        fun show(fragmentManager: FragmentManager) {
            GalleryImportPhotosDialogFragment().show(fragmentManager, TAG)
        }
    }

    private val viewModel: GallerySettingsViewModel by viewModels()
    private val getContentActivitiesLiveData: LiveData<List<ActivityInfo>> by lazy {
        viewModel.getContentActivityInfoList
    }
    private var listener: OnRequestContentListener? = null
    private lateinit var adapter: ArrayAdapter<CharSequence>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requireContext().withStyledAttributes(attrs = R.styleable.AlertDialog, defStyleAttr = R.attr.alertDialogStyle) {
            @LayoutRes val listItemLayout = getResourceId(R.styleable.AlertDialog_listItemLayout, 0)
            adapter = ArrayAdapter(requireContext(), listItemLayout)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        return MaterialAlertDialogBuilder(requireContext(),
                R.style.Theme_MaterialComponents_DayNight_Dialog_Alert)
                .setTitle(R.string.gallery_import_dialog_title)
                .setAdapter(adapter) { _, which ->
                    getContentActivitiesLiveData.value?.run {
                        listener?.requestGetContent(get(which))
                    }
                }.create()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = context as? OnRequestContentListener ?: throw IllegalArgumentException(
                "${context.javaClass.simpleName} must implement OnRequestContentListener")
        getContentActivitiesLiveData.observe(this) { getContentActivities ->
            if (getContentActivities.isEmpty()) {
                dismiss()
            } else {
                updateAdapter(getContentActivities)
            }
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    private fun updateAdapter(getContentActivites: List<ActivityInfo>) {
        val packageManager = requireContext().packageManager
        adapter.apply {
            clear()
            addAll(getContentActivites.map { it.loadLabel(packageManager) })
            notifyDataSetChanged()
        }
    }

    interface OnRequestContentListener {
        fun requestGetContent(info: ActivityInfo)
    }
}
